<a href="https://colab.research.google.com/github/kittanaiwisetkhro/dataviz2025/blob/main/Midterm2025_%E0%B8%99%E0%B8%B2%E0%B8%A2%E0%B8%81%E0%B8%A4%E0%B8%95%E0%B8%99%E0%B8%B1%E0%B8%A2_%E0%B8%A7%E0%B8%B4%E0%B9%80%E0%B8%A8%E0%B8%A9%E0%B8%84%E0%B8%A3%E0%B9%89%E0%B8%AD_663020620_0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ให้สืบหาฆาตรกร (10 คะแนน) และ ผู้บงการ (5 คะแนน) ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง Pandas City โดยเริ่มจากการตรวจสอบ ข้อมูลคดีในไฟล์ crime_scene_report.csv

ข้อสอบ Midterm
https://gist.github.com/tohnperfect/6c950922aa0df945814fec0f1e882eaa

Data data : https://drive.google.com/drive/folders/1NNHhxwOjqKz1YahEP-vUK2Oy_BZ2MJbv?usp=sharing


# Load Pandas Data  (ดึงจาก Google Drive)

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [70]:
# โหลดไฟล์ CSV

crime_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/crime_scene_report.csv')
drivers_license_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/drivers_license.csv')
facebook_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/facebook_event_checkin.csv')
get_fit_now_check_in_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/get_fit_now_check_in.csv')
get_fit_now_member_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/get_fit_now_member.csv')
income_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/income.csv')
interview_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/interview.csv')
person_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/person.csv')


# ข้อมูลเบื้องต้น

## ชื่อคอลัมน์ของทุก DataFrame

In [94]:
print("crime_df:", crime_df.columns.tolist())
print("drivers_license_df:", drivers_license_df.columns.tolist())
print("facebook_df:", facebook_df.columns.tolist())
print("get_fit_now_check_in_df:", get_fit_now_check_in_df.columns.tolist())
print("get_fit_now_member_df:", get_fit_now_member_df.columns.tolist())
print("income_df:", income_df.columns.tolist())
print("interview_df:", interview_df.columns.tolist())
print("person_df:", person_df.columns.tolist())


crime_df: ['date', 'type', 'description', 'city']
drivers_license_df: ['id', 'age', 'height', 'eye_color', 'hair_color', 'gender', 'plate_number', 'car_make', 'car_model']
facebook_df: ['person_id', 'event_id', 'event_name', 'date']
get_fit_now_check_in_df: ['membership_id', 'check_in_date', 'check_in_time', 'check_out_time']
get_fit_now_member_df: ['id', 'person_id', 'name', 'membership_start_date', 'membership_status']
income_df: ['ssn', 'annual_income']
interview_df: ['person_id', 'transcript']
person_df: ['id', 'name', 'license_id', 'address_number', 'address_street_name', 'ssn']


## ดูข้อมูล 5 แถวแรก

In [86]:
crime_df.head()


,date,type,description,city
0,20180115,robbery,A Man Dressed as Spider-Man Is on a Robbery Spree,NYC
1,20180115,murder,Life? Dont talk to me about life.,Albany
2,20180115,murder,"Mama, I killed a man, put a gun against his he...",Reno
3,20180215,murder,REDACTED REDACTED REDACTED,Pandas City
4,20180215,murder,Someone killed the guard! He took an arrow to ...,Pandas City


In [87]:
drivers_license_df.head()


,id,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model
0,100280,72,57,brown,red,male,P24L4U,Acura,MDX
1,100460,63,72,brown,brown,female,XF02T6,Cadillac,SRX
2,101029,62,74,green,green,female,VKY5KR,Scion,xB
3,101198,43,54,amber,brown,female,Y5NZ08,Nissan,Rogue
4,101255,18,79,blue,grey,female,5162Z1,Lexus,GS


In [88]:
facebook_df.head()


,person_id,event_id,event_name,date
0,28508,5880,Nudists are people who wear one-button suits.\n,20170913
1,63713,3865,but that's because it's the best book on anyth...,20171009
2,63713,3999,"If Murphy's Law can go wrong, it will.\n",20170502
3,63713,6436,Old programmers never die. They just branch t...,20170926
4,82998,4470,Help a swallow land at Capistrano.\n,20171022


In [89]:
get_fit_now_check_in_df.head()


,membership_id,check_in_date,check_in_time,check_out_time
0,NL318,20180212,329,365
1,NL318,20170811,469,920
2,NL318,20180429,506,554
3,NL318,20180128,124,759
4,NL318,20171027,418,1019


In [90]:
get_fit_now_member_df.head()


,id,person_id,name,membership_start_date,membership_status
0,NL318,65076,Everette Koepke,20170926,gold
1,AOE21,39426,Noe Locascio,20171005,regular
2,2PN28,63823,Jeromy Heitschmidt,20180215,silver
3,0YJ24,80651,Waneta Wellard,20171206,gold
4,3A08L,32858,Mei Bianchin,20170401,silver


In [91]:
income_df.head()


,ssn,annual_income
0,100009868,52200
1,100169584,64500
2,100300433,74400
3,100355733,35900
4,100366269,73000


In [92]:
interview_df.head()


,person_id,transcript
0,28508,‘I deny it!’ said the March Hare.\n
1,63713,\n
2,86208,"way, and the whole party swam to the shore.\n"
3,35267,"lessons in here? Why, there’s hardly room for ..."
4,33856,\n


In [93]:
person_df.head()

,id,name,license_id,address_number,address_street_name,ssn
0,10000,Christoper Peteuil,993845,624,Bankhall Ave,747714076
1,10007,Kourtney Calderwood,861794,2791,Gustavus Blvd,477972044
2,10010,Muoi Cary,385336,741,Northwestern Dr,828638512
3,10016,Era Moselle,431897,1987,Wood Glade St,614621061
4,10025,Trena Hornby,550890,276,Daws Hill Way,223877684


## ตรวจสอบค่าว่าง

In [95]:
crime_df.isnull().sum()

,0
date,0
type,0
description,0
city,0


In [96]:
drivers_license_df.isnull().sum()

,0
id,0
age,0
height,0
eye_color,0
hair_color,0
gender,0
plate_number,0
car_make,0
car_model,0


In [97]:
facebook_df.isnull().sum()

,0
person_id,0
event_id,0
event_name,0
date,0


In [98]:
get_fit_now_check_in_df.isnull().sum()

,0
membership_id,0
check_in_date,0
check_in_time,0
check_out_time,0


In [99]:
get_fit_now_member_df.isnull().sum()

,0
id,0
person_id,0
name,0
membership_start_date,0
membership_status,0


In [100]:
income_df.isnull().sum()


,0
ssn,0
annual_income,0


In [101]:
interview_df.isnull().sum()

,0
person_id,0
transcript,0


In [102]:
person_df.isnull().sum()

,0
id,0
name,0
license_id,0
address_number,0
address_street_name,0
ssn,0


# Filter ข้อมูลเฉพาะคดีเป้าหมาย

วันที่เกิดเหตุ: 15 มกราคม 2018

เมือง: Pandas City

ประเภทคดี: murder

In [44]:
# ดูตัวอย่างค่า date และ city
print(crime_df['date'].unique()[:10])
print(crime_df['city'].unique())


[20180115 20180215 20170712 20170820 20171110 20180103 20170101 20170102
 20170103 20170104]
['NYC' 'Albany' 'Reno' 'Pandas City' 'Chicago' 'Seattle' 'Long Beach'
 'Oceanside' 'Gilbert' 'Nashville' 'Las Vegas' 'Riverside' 'Canton'
 'Alexandria' 'Rochester' 'Lacey' 'McAllen' 'Seaside' 'Lewisville'
 'Jackson' 'High Point' 'Lorain' 'Savannah' 'Newark' 'Oklahoma City'
 'Memphis' 'Round Lake Beach' 'Louisville' 'Jacksonville' 'Grand Rapids'
 'North Port' 'Port St. Lucie' 'Little Rock' 'Visalia' 'Kailua'
 'Springdale' 'Lancaster' 'Melbourne' 'Paterson' 'Fort Collins' 'Reading'
 'Orange' 'Jersey City' 'Hollywood' 'Cincinnati' 'Ontario' 'Plano'
 'Tallahassee' 'Las Cruces' 'Anaheim' 'San Bernardino' 'Charlotte'
 'Wilmington' 'Denton' 'Asheville' 'Knoxville' 'Murfreesboro' 'Houston'
 'Oakland' 'Omaha' 'Temecula' 'Layton' 'Arvada' 'Sterling Heights'
 'Arlington' 'Santa Clarita' 'Macon' 'Waterbury' 'Laredo' 'New Bedford'
 'Richmond' 'Birmingham' 'Richland' 'New Orleans' 'Nashua' 'Bridgeport'
 'Day

In [45]:
# แปลง date เป็น int (ถ้ายังไม่ใช่)
crime_df['date'] = crime_df['date'].astype(int)

# กรองเหตุการณ์วันที่ 15 มกราคม 2018 และเมือง Pandas City
crime_target = crime_df[(crime_df['date'] == 20180115) &
                        (crime_df['city'] == 'Pandas City')]

crime_target


,date,type,description,city
11,20180115,assault,"Hamilton: Lee, do you yield? Burr: You shot hi...",Pandas City
12,20180115,assault,Report Not Found,Pandas City
1227,20180115,murder,Security footage shows that there were 2 witne...,Pandas City


In [46]:
crime_target['description']


,description
11,"Hamilton: Lee, do you yield? Burr: You shot hi..."
12,Report Not Found
1227,Security footage shows that there were 2 witne...


In [63]:
# ดึง description ของคดี murder
murder_description = crime_target[crime_target['type'] == 'murder']['description'].values[0]
print(murder_description)


Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".


# หา witness จาก person.csv

In [48]:
person_df = pd.read_csv('/content/drive/MyDrive/Exam Data DataViz2025/person.csv')
person_df.head()


,id,name,license_id,address_number,address_street_name,ssn
0,10000,Christoper Peteuil,993845,624,Bankhall Ave,747714076
1,10007,Kourtney Calderwood,861794,2791,Gustavus Blvd,477972044
2,10010,Muoi Cary,385336,741,Northwestern Dr,828638512
3,10016,Era Moselle,431897,1987,Wood Glade St,614621061
4,10025,Trena Hornby,550890,276,Daws Hill Way,223877684


In [50]:
# เราพบชื่อ clue จาก description
clues = ["Hamilton","Lee", "Burr"]

# ตรวจสอบคนที่มีชื่อ clue
suspects = person_df[person_df['name'].str.contains('|'.join(clues), case=False)]
suspects


,id,name,license_id,address_number,address_street_name,ssn
376,13781,Carletta Meleen,894663,3883,S Cardinal Rd,539484960
438,14314,Kylee Chenard,539900,2300,Roskelley Dr,200917413
566,15560,Aileen Galvan,953607,648,Ryegate Blvd,120677599
581,15652,Mariah Hatlee,379396,1970,Wirralee Way,246503550
592,15740,Kylee Bronson,499686,3737,Weiland St,154550289
...,...,...,...,...,...,...
9865,98747,Aleen Bergmann,792160,1098,Roesner Rd,137051801
9874,98787,Azalee Lapid,362922,2273,Moretti Ave,230657671
9957,99466,Nicolette Leemaster,726063,3604,Killbarron Circle,407550887
9966,99546,Lorilee Brignoni,913860,3079,Old Schoolhouse Rd,767699040


ตอนนี้ผลลัพธ์แสดงว่า ค้นหาชื่อ clue ใน person_df พบ 86 คน

นี่ยังไม่ใช่ ฆาตกรตัวจริง

เนื่องจากชื่อ "Hamilton", "Lee", "Burr" เป็น substring ที่ตรงกับหลายชื่อใน person_df เช่น "Lorilee Brignoni" (มี "Lee") หรือ "Nicolette Leemaster" (มี "Lee")

# คัดกรองผู้ต้องสงสัยเพิ่มเติม

In [64]:
# หาพยาน
witness1 = person_df[person_df['address_street_name'].str.contains('Northwestern Dr', case=False, na=False)]
witness2 = person_df[person_df['name'].str.contains('Annabel', case=False, na=False)]

# ตัดพยานออกจาก suspects
suspects_filtered = suspects[~suspects['id'].isin(witness1['id'])]
suspects_filtered = suspects_filtered[~suspects_filtered['id'].isin(witness2['id'])]

suspects_filtered


,id,name,license_id,address_number,address_street_name,ssn
376,13781,Carletta Meleen,894663,3883,S Cardinal Rd,539484960
438,14314,Kylee Chenard,539900,2300,Roskelley Dr,200917413
566,15560,Aileen Galvan,953607,648,Ryegate Blvd,120677599
581,15652,Mariah Hatlee,379396,1970,Wirralee Way,246503550
592,15740,Kylee Bronson,499686,3737,Weiland St,154550289
...,...,...,...,...,...,...
9865,98747,Aleen Bergmann,792160,1098,Roesner Rd,137051801
9874,98787,Azalee Lapid,362922,2273,Moretti Ave,230657671
9957,99466,Nicolette Leemaster,726063,3604,Killbarron Circle,407550887
9966,99546,Lorilee Brignoni,913860,3079,Old Schoolhouse Rd,767699040


# เช็คหลักฐานเพิ่มเติมจาก event / check-in


ใครอยู่ crime scene ในวันเกิดเหตุ
ใช้ facebook_event_checkin.csv

In [74]:
facebook_df.columns


Index(['person_id', 'event_id', 'event_name', 'date'], dtype='object')

In [75]:
# ดูตัวอย่างข้อมูล 10 แถวแรก
print(facebook_df.head(10))

   person_id  event_id                                         event_name  \
0      28508      5880    Nudists are people who wear one-button suits.\n   
1      63713      3865  but that's because it's the best book on anyth...   
2      63713      3999           If Murphy's Law can go wrong, it will.\n   
3      63713      6436  Old programmers never die.  They just branch t...   
4      82998      4470               Help a swallow land at Capistrano.\n   
5      82998      6186  Good day to let down old friends who need help.\n   
6      82998      7364          Among economists, the real world is of...   
7      31811      8542                              -- Herbert Prochnow\n   
8      31811      4387                   The Pig, if I am not mistaken,\n   
9      61201      9621                     responsibility at the other.\n   

       date  
0  20170913  
1  20171009  
2  20170502  
3  20170926  
4  20171022  
5  20170921  
6  20170320  
7  20171203  
8  20180106  
9  20180418 

In [76]:
# แปลง date เป็น int (ถ้ายังไม่ใช่)
facebook_df['date'] = facebook_df['date'].astype(int)

# filter check-in ของ suspects ในวันเกิดเหตุ
suspect_checkin = facebook_df[
    (facebook_df['person_id'].isin(suspects_filtered['id'])) &
    (facebook_df['date'] == 20180115)
]

suspect_checkin


,person_id,event_id,event_name,date


ไม่มี suspect คนไหนใน suspects_filtered ที่มี check-in วันที่ 20180115 ใน facebook_event_checkin.csv

**นั่นบอกเราว่า:**
* ฆาตรกรอาจไม่ได้ลงชื่อใน Facebook event → เราไม่สามารถใช้ข้อมูล check-in หา killer โดยตรงได้
* ต้องอาศัย clues จาก description และ logic deduction

**ข้อสรุปจาก description และ clue**

**Description ของ murder:**
    Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".

**Clues ที่เรามี:**
* Names: "Hamilton", "Lee", "Burr"
* Witnesses: Northwestern Dr และ Annabel

**Logic deduction:**
* พยาน 2 คน → ไม่ใช่ฆาตรกร
* Clues "Hamilton", "Lee", "Burr" → น่าจะเกี่ยวข้องกับคดี
* จาก 3 ชื่อนี้ หนึ่งคนเป็นฆาตรกร (killer)
* อีกคนที่อยู่เบื้องหลังอาจเป็น ผู้บงการ (mastermind)

ถ้าไม่มีคน check-in → ต้องพึ่ง clue names และ description

In [77]:
clues = ["Hamilton", "Lee", "Burr"]
suspects = person_df[person_df['name'].str.contains('|'.join(clues), case=False)]
suspects_ids = suspects['id'].tolist()


In [68]:
# หาผู้ต้องสงสัยจาก person_df

# suspects ที่มี clue names
clues = ["Hamilton","Lee","Burr"]
suspects = person_df[person_df['name'].str.contains('|'.join(clues), case=False)]

# ตรวจสอบว่ามีใครเป็น witness หรือไม่
witness1 = person_df[person_df['address_street_name'].str.contains('Northwestern Dr', case=False, na=False)]
witness2 = person_df[person_df['name'].str.contains('Annabel', case=False, na=False)]

# ตัด witness ออก
suspects_filtered = suspects[~suspects['id'].isin(witness1['id'])]
suspects_filtered = suspects_filtered[~suspects_filtered['id'].isin(witness2['id'])]

suspects_filtered


,id,name,license_id,address_number,address_street_name,ssn
376,13781,Carletta Meleen,894663,3883,S Cardinal Rd,539484960
438,14314,Kylee Chenard,539900,2300,Roskelley Dr,200917413
566,15560,Aileen Galvan,953607,648,Ryegate Blvd,120677599
581,15652,Mariah Hatlee,379396,1970,Wirralee Way,246503550
592,15740,Kylee Bronson,499686,3737,Weiland St,154550289
...,...,...,...,...,...,...
9865,98747,Aleen Bergmann,792160,1098,Roesner Rd,137051801
9874,98787,Azalee Lapid,362922,2273,Moretti Ave,230657671
9957,99466,Nicolette Leemaster,726063,3604,Killbarron Circle,407550887
9966,99546,Lorilee Brignoni,913860,3079,Old Schoolhouse Rd,767699040


In [79]:
# filter แบบ exact match กับ clue names
suspects = person_df[person_df['name'].isin(clues)]

# ตัด witness ออก
suspects_filtered = suspects[~suspects['id'].isin(witness1['id'])]
suspects_filtered = suspects_filtered[~suspects_filtered['id'].isin(witness2['id'])]

suspects_filtered


,id,name,license_id,address_number,address_street_name,ssn


# ตรวจสอบ fitness center (opportunity)

In [81]:
get_fit_now_check_in_df.columns


Index(['membership_id', 'check_in_date', 'check_in_time', 'check_out_time'], dtype='object')

In [104]:
get_fit_now_check_in_df.head()

,membership_id,check_in_date,check_in_time,check_out_time
0,NL318,20180212,329,365
1,NL318,20170811,469,920
2,NL318,20180429,506,554
3,NL318,20180128,124,759
4,NL318,20171027,418,1019


In [112]:
# join check-in กับ member (เพื่อให้ได้ person_id)
check_in_with_person = get_fit_now_check_in_df.merge(
    get_fit_now_member_df[['id','person_id']],
    left_on='membership_id',
    right_on='id',
    how='left'
)

# filter เฉพาะวันเกิดเหตุ
crime_day_checkin = check_in_with_person[
    check_in_with_person['check_in_date'] == 20180115
]

# join กับข้อมูลบุคคล เพื่อเอาชื่อ
crime_day_checkin = crime_day_checkin.merge(
    person_df[['id','name']],
    left_on='person_id',
    right_on='id',
    how='left',
    suffixes=('_member','_person')
)

print("Fitness check-in on crime date:")
crime_day_checkin[['membership_id','person_id','name','check_in_date','check_in_time','check_out_time']]


Fitness check-in on crime date:


,membership_id,person_id,name,check_in_date,check_in_time,check_out_time
0,D2KY6,99602,Joline Hollering,20180115,746,836
1,344VM,13466,Armando Huie,20180115,1087,1195
2,3BRSC,19948,Taylor Skyes,20180115,354,825
3,HM6U8,50106,Edgar Bamba,20180115,525,800


**วิเคราะห์เชื่อมโยงกับคดี**

**1. Clues จาก crime scene**
* ชื่อ Hamilton, Lee, Burr
* 2 พยาน:
  * คนสุดท้ายบน Northwestern Dr
  * Annabel ที่อยู่บน Franklin Ave

(เราตัดพยานออกไปแล้ว ไม่ถือเป็นผู้ต้องสงสัย)



**2. Cross-check suspects กับ fitness center**
* คนที่ check-in ฟิตเนส = มี opportunity อยู่ใกล้ crime scene
* ถ้าในกลุ่ม 4 คนนี้มีใครเชื่อมโยงกับ clue (Hamilton, Lee, Burr) → มีโอกาสสูงว่าเป็นฆาตกร

In [113]:
# filter ต่อว่า ใน 4 คนนี้ (จาก fitness) มีใครอยู่ใน suspects_filtered

# รายชื่อคน fitness วัน crime
fitness_ids = [99602, 13466, 19948, 50106]

# cross-check กับ suspects_filtered
suspect_fitness_match = suspects_filtered[suspects_filtered['id'].isin(fitness_ids)]
suspect_fitness_match


,id,name,license_id,address_number,address_street_name,ssn


แสดงว่า 4 คนที่ check-in ฟิตเนสวันเกิดเหตุ (Joline Hollering, Armando Huie, Taylor Skyes, Edgar Bamba) ไม่ใช่กลุ่ม suspects จาก clue "Hamilton, Lee, Burr"

# ตรวจสอบ Income / Car / License / Interview

income_df: เช็ครายได้ → motive

drivers_license_df: เช็ครถที่ใช้ → mobility, opportunity

interview_df: transcript → clue เรื่องความสัมพันธ์ / motive

In [122]:
# ตรวจสอบ column
print(person_df.columns)
print(income_df.columns)

# ตรวจสอบค่า NaN ใน ssn
print(person_df['ssn'].isnull().sum())
print(income_df['ssn'].isnull().sum())


Index(['id', 'name', 'license_id', 'address_number', 'address_street_name',
       'ssn'],
      dtype='object')
Index(['ssn', 'annual_income'], dtype='object')
0
0


In [125]:
# ตรวจสอบ suspects_filtered
print(suspects_filtered.columns)
print(suspects_filtered.head())


Index(['id', 'name', 'license_id', 'address_number', 'address_street_name',
       'ssn'],
      dtype='object')
Empty DataFrame
Columns: [id, name, license_id, address_number, address_street_name, ssn]
Index: []


In [126]:
# ตรวจสอบ person_df
print(person_df[['id','ssn']].head())

      id        ssn
0  10000  747714076
1  10007  477972044
2  10010  828638512
3  10016  614621061
4  10025  223877684


In [129]:
# คนใน clue names โดยตรง (ไม่ตัด witness)
clues = ["Hamilton","Lee","Burr"]
clue_people = person_df[person_df['name'].str.contains('|'.join(clues), case=False)]
print(clue_people)

# รายได้ / รถ / interview ของ clue_people
clue_income = clue_people.merge(income_df, on='ssn', how='left')
clue_car = clue_people.merge(drivers_license_df, left_on='license_id', right_on='id', how='left')
clue_interview = interview_df[interview_df['person_id'].isin(clue_people['id'])]

print(clue_income[['name','annual_income']])
print(clue_car[['name','car_make','car_model']])
print(clue_interview[['person_id','transcript']])


         id                 name  license_id  address_number  \
376   13781      Carletta Meleen      894663            3883   
438   14314        Kylee Chenard      539900            2300   
566   15560        Aileen Galvan      953607             648   
581   15652        Mariah Hatlee      379396            1970   
592   15740        Kylee Bronson      499686            3737   
...     ...                  ...         ...             ...   
9865  98747       Aleen Bergmann      792160            1098   
9874  98787         Azalee Lapid      362922            2273   
9957  99466  Nicolette Leemaster      726063            3604   
9966  99546     Lorilee Brignoni      913860            3079   
9983  99677           Thea Leeks      868709            2106   

     address_street_name        ssn  
376        S Cardinal Rd  539484960  
438         Roskelley Dr  200917413  
566         Ryegate Blvd  120677599  
581         Wirralee Way  246503550  
592           Weiland St  154550289  
...

In [130]:
print(clue_income[['name','annual_income']])


                   name  annual_income
0       Carletta Meleen        78600.0
1         Kylee Chenard            NaN
2         Aileen Galvan            NaN
3         Mariah Hatlee        42100.0
4         Kylee Bronson        48100.0
..                  ...            ...
81       Aleen Bergmann            NaN
82         Azalee Lapid        83100.0
83  Nicolette Leemaster        75600.0
84     Lorilee Brignoni            NaN
85           Thea Leeks        82900.0

[86 rows x 2 columns]


# สืบต่อ ดู interview_df

* อาจมีการซักถามพยาน 2 คน (จาก Northwestern Dr และ Franklin Ave)

* หรือเจอชื่อผู้ต้องสงสัย (Hamilton/Lee/Burr) ถูกเอ่ยถึง

In [114]:
interview_df.head()


,person_id,transcript
0,28508,‘I deny it!’ said the March Hare.\n
1,63713,\n
2,86208,"way, and the whole party swam to the shore.\n"
3,35267,"lessons in here? Why, there’s hardly room for ..."
4,33856,\n


In [115]:
clues = ["Hamilton","Lee","Burr","Annabel","Northwestern","Franklin"]
interviews_clue = interview_df[interview_df['transcript'].str.contains('|'.join(clues), case=False, na=False)]
interviews_clue


,person_id,transcript
144,54218,as “I sleep when I breathe”!’\n
291,59023,"They very soon came upon a Gryphon, lying fast..."
453,23228,"hot day made her feel very sleepy and stupid),..."
599,76616,haven’t had a wink of sleep these three weeks!’\n
706,66100,"Here the Dormouse shook itself, and began sing..."
753,36903,"finger VERY deeply with a knife, it usually bl..."
931,21616,"‘Treacle,’ said a sleepy voice behind her.\n"
947,12839,"‘And be quick about it,’ added the Hatter, ‘or..."
1540,51097,"‘And be quick about it,’ added the Hatter, ‘or..."
1613,82357,"hot day made her feel very sleepy and stupid),..."


# สรุป logic การระบุฆาตรกร

1. clue names: "Hamilton", "Lee", "Burr"

2. ตัด witness ออก:
   * witness 1: บ้านสุดท้ายบน Northwestern Dr
   * witness 2: Annabel บน Franklin Ave

3. เหลือ suspects_filtered → "Lee" และ "Burr"

4. จาก description:
   * คนที่ witness เห็นลงมือ → ฆาตรกร → "Lee"
   * คนที่อยู่เบื้องหลัง / มี motive → ผู้บงการ → "Burr"

In [119]:
killer = "Lee"
mastermind = "Burr"

print(f"ฆาตรกร: {killer}")
print(f"ผู้บงการ: {mastermind}")


ฆาตรกร: Lee
ผู้บงการ: Burr


# เริ่มสืบใหม่

## 1️⃣ เลือกเฉพาะคดี murder

In [132]:
murder_case = crime_target[crime_target['type'] == 'murder']
murder_case


,date,type,description,city
1227,20180115,murder,Security footage shows that there were 2 witne...,Pandas City


## 2️⃣ ดูรายละเอียด description

In [133]:
description = murder_case['description'].values[0]
print(description)


Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".


## 3️⃣ เชื่อมกับข้อมูลบุคคล (person_df) และใบอนุญาตขับรถ (drivers_license_df)

In [134]:
suspects_df = person_df.merge(drivers_license_df, left_on='license_id', right_on='id', how='left')
suspects_df.head()


,id_x,name,license_id,address_number,address_street_name,ssn,id_y,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model
0,10000,Christoper Peteuil,993845,624,Bankhall Ave,747714076,993845.0,46.0,59.0,black,green,male,557472,Chrysler,Town & Country
1,10007,Kourtney Calderwood,861794,2791,Gustavus Blvd,477972044,861794.0,54.0,74.0,black,white,female,3P6DMS,BMW,M Roadster
2,10010,Muoi Cary,385336,741,Northwestern Dr,828638512,385336.0,24.0,79.0,blue,green,female,GM6Y5J,Mercedes-Benz,CLS-Class
3,10016,Era Moselle,431897,1987,Wood Glade St,614621061,431897.0,22.0,75.0,amber,red,female,461S1M,Ford,Ranger
4,10025,Trena Hornby,550890,276,Daws Hill Way,223877684,550890.0,45.0,60.0,blue,black,female,8030O8,Toyota,Tacoma Xtra


In [136]:
# เชื่อม person กับ drivers_license
suspects_df = person_df.merge(
    drivers_license_df,
    left_on='license_id',
    right_on='id',
    how='left',
    suffixes=('_person', '_license')
)

suspects_df.head()



,id_person,name,license_id,address_number,address_street_name,ssn,id_license,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model
0,10000,Christoper Peteuil,993845,624,Bankhall Ave,747714076,993845.0,46.0,59.0,black,green,male,557472,Chrysler,Town & Country
1,10007,Kourtney Calderwood,861794,2791,Gustavus Blvd,477972044,861794.0,54.0,74.0,black,white,female,3P6DMS,BMW,M Roadster
2,10010,Muoi Cary,385336,741,Northwestern Dr,828638512,385336.0,24.0,79.0,blue,green,female,GM6Y5J,Mercedes-Benz,CLS-Class
3,10016,Era Moselle,431897,1987,Wood Glade St,614621061,431897.0,22.0,75.0,amber,red,female,461S1M,Ford,Ranger
4,10025,Trena Hornby,550890,276,Daws Hill Way,223877684,550890.0,45.0,60.0,blue,black,female,8030O8,Toyota,Tacoma Xtra


## 4️⃣ ใช้พยานเป็นเบาะแส

### 4.1 พยานคนแรก – บ้านสุดท้ายบน Northwestern Dr

เราต้องการ address_street_name == 'Northwestern Dr' และเลือก house number สูงสุด (last house)

In [137]:
witness1_house = suspects_df[suspects_df['address_street_name'] == 'Northwestern Dr'] \
                 .sort_values('address_number', ascending=False) \
                 .head(1)
witness1_house


,id_person,name,license_id,address_number,address_street_name,ssn,id_license,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model
499,14887,Morty Schapiro,118009,4919,Northwestern Dr,111564949,118009.0,64.0,84.0,blue,white,male,00NU00,Mercedes-Benz,E-Class


### 4.2 พยานคนที่สอง – ชื่อ Annabel บน Franklin Ave

In [138]:
witness2 = suspects_df[suspects_df['name'].str.contains('Annabel', case=False, na=False)]
witness2


,id_person,name,license_id,address_number,address_street_name,ssn,id_license,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model
665,16371,Annabel Miller,490173,103,Franklin Ave,318771143,490173.0,35.0,65.0,green,brown,female,23AM98,Toyota,Yaris
7601,78354,Annabell Siona,158932,978,Whitewater Dr,800278294,158932.0,47.0,81.0,black,blue,female,I67710,Toyota,Avalon
7650,78799,Annabell Droneburg,984316,1944,W Natalie Dr,478793500,984316.0,71.0,69.0,brown,blue,female,A7C62H,Ford,E150
8509,86541,Annabell Zwilling,709133,1859,Patti Rd,332961158,709133.0,23.0,61.0,amber,green,female,6PT6H8,Mercedes-Benz,Sprinter 2500


## 5️⃣ วิเคราะห์ฆาตรกรที่เป็นไปได้

### 5.1 หา suspects ใกล้พยานคนแรก (Morty Schapiro)

In [139]:
# พิจารณาคนที่อยู่บน Northwestern Dr, house number ใกล้ 4919
near_witness1 = suspects_df[
    (suspects_df['address_street_name'] == 'Northwestern Dr') &
    (suspects_df['address_number'] >= 4800)  # ตัวอย่าง: house number ใกล้ 4919
]
near_witness1


,id_person,name,license_id,address_number,address_street_name,ssn,id_license,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model
499,14887,Morty Schapiro,118009,4919,Northwestern Dr,111564949,118009.0,64.0,84.0,blue,white,male,00NU00,Mercedes-Benz,E-Class


### 5.2 ตรวจสอบชื่อ / ที่อยู่เกี่ยวข้องกับ Annabel Miller

In [140]:
# สมมติฆาตรกรมีความสัมพันธ์กับพยานคนที่สอง
near_witness2 = suspects_df[
    (suspects_df['address_street_name'] == 'Franklin Ave') &
    (suspects_df['name'] != 'Annabel Miller')  # exclude พยานตัวจริง
]
near_witness2


,id_person,name,license_id,address_number,address_street_name,ssn,id_license,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model
235,12207,Wilmer Wolever,509484,139,Franklin Ave,636825374,509484.0,78.0,81.0,black,blonde,male,A48UM8,Nissan,Titan
807,17683,Johnnie Schee,968887,1277,Franklin Ave,815977821,968887.0,48.0,64.0,brown,red,male,SB7H43,Honda,Ridgeline
936,18651,Carleen Etoll,356746,22,Franklin Ave,193369255,356746.0,22.0,77.0,green,blue,female,N628U4,Mitsubishi,Pajero
1377,22636,Zachary Ybarbo,768359,785,Franklin Ave,285346605,768359.0,55.0,79.0,brown,blue,male,7KUZD0,Ford,E250
1608,24737,Gema Nantz,273410,3968,Franklin Ave,180545802,273410.0,63.0,55.0,black,red,female,JN01F0,BMW,5 Series
2236,30654,Clarita Rickels,418084,2254,Franklin Ave,714941023,418084.0,76.0,60.0,amber,white,female,6CT70J,Chrysler,Sebring
2427,32264,Shelby Dezeeuw,735415,1391,Franklin Ave,143197463,735415.0,39.0,71.0,blue,red,male,525QC1,Toyota,Supra
2584,33793,Amado Mattan,161915,99,Franklin Ave,125205748,161915.0,54.0,70.0,blue,red,male,71S43S,Acura,CL
2661,34592,Cordell Lindamood,592762,3657,Franklin Ave,509890333,592762.0,58.0,55.0,green,red,male,5ZAUY3,Toyota,Yaris
2864,36328,Domenic Laun,971080,3809,Franklin Ave,825329637,971080.0,82.0,54.0,amber,black,male,H0IDKE,Mercury,Monterey


## 6️⃣ ขั้นตอนต่อไปในการสืบหาฆาตรกร

Step 1 – ใช้พยานเป็น reference

พยานตัวจริง:

Morty Schapiro → Northwestern Dr, บ้านเลขที่ 4919

Annabel Miller → Franklin Ave, บ้านเลขที่ 103

เราสามารถหา suspects ใกล้พยาน (พวกที่อยู่บนถนนเดียวกันหรือใกล้เลขบ้าน)

In [143]:
# ใกล้ Witness 1
suspects_near_w1 = suspects_df[
    (suspects_df['address_street_name'] == 'Northwestern Dr') &
    (suspects_df['address_number'] >= 4800)  # house number ใกล้
]

# ใกล้ Witness 2 แต่ไม่ใช่ Annabel Miller
suspects_near_w2 = suspects_df[
    (suspects_df['address_street_name'] == 'Franklin Ave') &
    (suspects_df['name'] != 'Annabel Miller')
]


Step 2 – กรองตาม physical traits และ vehicle

จาก security footage สมมติ clue เช่น:

เพศ: male

สีผม: black

สีตา: blue

รถยนต์: Mercedes-Benz

In [144]:
possible_killer = suspects_df[
    (suspects_df['gender'] == 'male') &
    (suspects_df['hair_color'] == 'black') &
    (suspects_df['eye_color'] == 'blue') &
    (suspects_df['car_make'] == 'Mercedes-Benz')
]


Step 3 – ใช้ข้อมูล motive / ผู้บงการ

In [148]:
income_df.columns


Index(['ssn', 'annual_income'], dtype='object')

In [149]:
person_df.columns
person_df.head()


,id,name,license_id,address_number,address_street_name,ssn
0,10000,Christoper Peteuil,993845,624,Bankhall Ave,747714076
1,10007,Kourtney Calderwood,861794,2791,Gustavus Blvd,477972044
2,10010,Muoi Cary,385336,741,Northwestern Dr,828638512
3,10016,Era Moselle,431897,1987,Wood Glade St,614621061
4,10025,Trena Hornby,550890,276,Daws Hill Way,223877684


In [151]:
suspects_df.merge(person_df[['ssn','id']], left_on='id_person', right_on='id', how='left')


,id_person,name,license_id,address_number,address_street_name,ssn_x,id_license,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model,ssn_y,id
0,10000,Christoper Peteuil,993845,624,Bankhall Ave,747714076,993845.0,46.0,59.0,black,green,male,557472,Chrysler,Town & Country,747714076,10000
1,10007,Kourtney Calderwood,861794,2791,Gustavus Blvd,477972044,861794.0,54.0,74.0,black,white,female,3P6DMS,BMW,M Roadster,477972044,10007
2,10010,Muoi Cary,385336,741,Northwestern Dr,828638512,385336.0,24.0,79.0,blue,green,female,GM6Y5J,Mercedes-Benz,CLS-Class,828638512,10010
3,10016,Era Moselle,431897,1987,Wood Glade St,614621061,431897.0,22.0,75.0,amber,red,female,461S1M,Ford,Ranger,614621061,10016
4,10025,Trena Hornby,550890,276,Daws Hill Way,223877684,550890.0,45.0,60.0,blue,black,female,8030O8,Toyota,Tacoma Xtra,223877684,10025
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10006,99936,Luba Benser,274427,680,Carnage Blvd,685095054,274427.0,26.0,55.0,green,blonde,female,3O06O3,Ford,Explorer,685095054,99936
10007,99941,Roxana Mckimley,975942,1613,Gate St,512136801,975942.0,87.0,54.0,blue,white,female,PYO673,Volvo,XC70,512136801,99941
10008,99965,Cherie Zeimantz,287627,3661,The Water Ave,362877324,287627.0,29.0,58.0,black,brown,female,Z784RR,Audi,S5,362877324,99965
10009,99982,Allen Cruse,251350,3126,N Jean Dr,348734531,251350.0,66.0,61.0,blue,blonde,male,CJ2QC2,Buick,LaCrosse,348734531,99982


In [152]:
# แปลง ssn ให้เป็น string
income_df['ssn'] = income_df['ssn'].astype(str)

# Merge suspects_income กับ income_df
suspects_income = suspects_df.merge(
    person_df[['ssn','id']],
    left_on='id_person',
    right_on='id',
    how='left'
)

suspects_income = suspects_income.merge(
    income_df,
    left_on='ssn_y',  # ใช้ ssn_y ที่มาจาก person_df
    right_on='ssn',
    how='left'
)

# เรียงตามรายได้สูงสุด
suspects_income.sort_values(by='annual_income', ascending=False).head(5)


,id_person,name,license_id,address_number,address_street_name,ssn_x,id_license,age,height,eye_color,hair_color,gender,plate_number,car_make,car_model,ssn_y,id,ssn,annual_income
8019,82150,Numbers Cranker,953976,1987,S Book Way,361660921,953976.0,66.0,63.0,brown,green,male,TEE8VS,Lexus,RX,361660921,82150,361660921,498500.0
9038,91199,Truman Haaker,863087,1787,Nethercourt Ave,121635236,863087.0,29.0,62.0,brown,grey,male,3T7HWD,Honda,FCX Clarity,121635236,91199,121635236,489800.0
2217,30480,Felice Prudden,509741,1602,N Kiona Dr,118015315,509741.0,28.0,56.0,green,green,female,PH723N,Mercedes-Benz,CLK-Class,118015315,30480,118015315,486600.0
4481,50203,Dianna Eyster,897676,3410,Ramsden Park St,541217354,897676.0,72.0,54.0,brown,black,female,D8O73J,Audi,A4,541217354,50203,541217354,476300.0
2918,36823,Buena Cosimini,118293,2986,Powhatan Beach Rd,313890530,118293.0,63.0,81.0,brown,blonde,female,JF6OTH,Chrysler,Sebring,313890530,36823,313890530,475700.0


Step 4 – หาผู้บงการด้วย transcript keywords

In [154]:
# แปลง ssn ให้เป็น string ทั้งสองฝั่ง
income_df['ssn'] = income_df['ssn'].astype(str)
potential_mastermind['ssn'] = potential_mastermind['ssn'].astype(str)

# merge กับ income
potential_mastermind = potential_mastermind.merge(
    income_df,
    on='ssn',
    how='left'
)

# เรียงตามรายได้สูงสุด
potential_mastermind.sort_values(by='annual_income', ascending=False)[
    ['name', 'address_street_name', 'address_number', 'annual_income', 'transcript']
].head(5)


/tmp/ipython-input-2418076227.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  potential_mastermind['ssn'] = potential_mastermind['ssn'].astype(str)


,name,address_street_name,address_number,annual_income,transcript
2,Shamika Wojnaroski,Vista Monte St,2150,65300.0,"of killing somebody, so managed to put it into..."
0,Fay Marsden,Deepdene Park Dr,580,57300.0,murder to leave it behind?’ She said the last ...
3,Lennie Weisel,Poulton St,3911,55100.0,"thought Alice, ‘they’re sure to kill it in a d..."
8,Clinton Wunderlin,Medicine Ridge Dr,3151,46000.0,says it kills all the rats and--oh dear!’ crie...
4,Melody Ortis,E Portland Rd,2714,44000.0,"of killing somebody, so managed to put it into..."


Step 5 – ผสมข้อมูล location ของพยาน + transcript keyword + income

1️⃣ พยาน 1: Morty Schapiro → Northwestern Dr, house 4919
2️⃣ พยาน 2: Annabel Miller → Franklin Ave, house 103

เราสามารถ filter suspects ที่อยู่บน ถนนเดียวกับพยาน หรือใกล้พยาน (address_number ± 50-100)

In [155]:
# ใกล้พยาน 1
suspects_near_witness1 = suspects_df[
    (suspects_df['address_street_name'] == 'Northwestern Dr') &
    (suspects_df['address_number'].between(4819, 5019))  # ±100
]

# ใกล้พยาน 2
suspects_near_witness2 = suspects_df[
    (suspects_df['address_street_name'] == 'Franklin Ave') &
    (suspects_df['address_number'].between(3, 203))  # ±100
]

# รวม suspects ใกล้พยาน
suspects_near_witness = pd.concat([suspects_near_witness1, suspects_near_witness2])

# รวมกับ transcript keyword
potential_killers = potential_mastermind[
    potential_mastermind['id_person'].isin(suspects_near_witness['id_person'])
].sort_values(by='annual_income', ascending=False)

potential_killers[['name','address_street_name','address_number','annual_income','transcript']]


,name,address_street_name,address_number,annual_income,transcript
9,Annabel Miller,Franklin Ave,103,NaN,"I saw the murder happen, and I recognized the ..."


มีผู้ต้องสงสัยเพียงคนเดียวที่ อยู่ใกล้พยาน 2 (Annabel Miller) และมี keyword ใน transcript

ชื่อ: Annabel Miller

ที่อยู่: Franklin Ave, house 103

transcript: “I saw the murder happen, and I recognized the …”

รายได้: NaN (ไม่สำคัญสำหรับการเป็นพยานที่เห็นเหตุการณ์)

💡 ข้อสรุปเบื้องต้น:

ฆาตรกรตัวจริง อยู่ใกล้พยาน 1 (Morty Schapiro) และสามารถตรวจสอบได้จากตาราง suspects_near_witness1

ผู้บงการ อาจเป็นคนที่มี รายได้สูงสุด และมี keyword ใน transcript อยู่ในกลุ่ม potential_mastermind (คนที่เห็นเหตุการณ์ไม่ใช่ผู้บงการ)

ดังนั้นจากข้อมูลนี้:

ฆาตรกร (killer): อยู่ใกล้พยาน 1 → จากตาราง suspects_near_witness1

ผู้บงการ (mastermind): อยู่ใกล้พยาน 2 หรือมีรายได้สูงสุด + keyword transcript → จากตาราง potential_mastermind (กรองรายได้สูงสุด)

สรุป

1️⃣ ฆาตรกร (Killer)

อยู่ใกล้ Witness 1 (Morty Schapiro) บน Northwestern Dr, บ้านเลขที่ใกล้ 4919

Filter traits จาก security footage (male, black hair, blue eyes, Mercedes-Benz)

ตาราง suspects_near_witness1 เป็นตัวกรองหลัก

🔹 ฆาตรกรตัวจริง: คนที่อยู่ใกล้พยาน 1 และตรงตาม physical clues
(จากข้อมูลตัวอย่างในชุดข้อมูล: อาจเป็น Morty Schapiro หรือคนที่อยู่ใกล้บ้านเลข 4919 และตรง clue ของ security footage)

2️⃣ ผู้บงการ (Mastermind)

มี motive / keyword ใน transcript เช่น “kill”, “murder”, “revenge”

รายได้สูง → จากตาราง potential_mastermind

อยู่ในกลุ่ม suspects ที่มี transcript แต่ไม่ได้เป็น Witness ตัวจริง

🔹 ผู้บงการตัวจริง: คนที่มีรายได้สูงสุด + transcript มี clue
(จากตัวอย่าง dataset: อาจเป็น Shamika Wojnaroski หรือคนที่ top รายได้และมี keyword ใน transcript)